working/4200-metaobjects/feature-specification.md

```diff
@@ -0,0 +1,519 @@
+it works exactly the same as today. So does the result of evaluating a type
+that does not denote a declaration (that is, a function type, a record
+type, special types like `dynamic`, union types like `T?`  and
+`FutureOr<T>`, etc.)
+
+The previous example showed how we can use metaobjects to provide access to
+static members of a set of classes (or mixins, etc.) without depending on
+the exact class (mixin, etc.) declaration. This basically means that we
+have made the static members _virtual_, because we're calling them via
+normal instance members (that is, virtual members, with all the usual OO
+semantics) of the metaobject.
+
+The next example illustrates how we can use metaobjects to call
+constructors in a similar way (yielding 'virtual constructors'):
+
+```dart
+abstract class Creation<X> {
+  const Creation();
+  X call();
+  X named(int _);
+}
+
+class C<Y> static implements Creation<C<Y>> {
+  C();
+  C.named(int _): this();
+}
+
+class D static implements Creation<D> {
+  factory D() = _DImpl;
+  D.named(int _);
+}
+
+class _DImpl implements D {...}
+
+X create<X static extends Creation<X>>() => X();
+
+void main() {
+  C<int> c = create();
+  D d = create();
+}
+```
+
+This illustrates that we can perform creation of instances of the given
+type argument (denoted by `X` in the declaration of `create`), in spite of
+the fact that the class `C` is generic (and the type argument `X` has the
+value `C<int>`, that is, it carries the actual type arguments with it), and
+the constructor in `C` that we're using is the implicitly induced 'default'
+constructor whose name is `C` (we could have written this explicitly as
+`C();` or `C(): super();`). In contrast, the constructor that we're using
+in `D` is a redirecting factory constructor.
+
+This works because the only requirement for a class `C` to static implement
+`Creation<C>` is that it must have a declaration which can be invoked as an
+invocation of the type itself (`C()`), which is exactly what we get the
+ability to do when the metaobject has a `call` method (that is, we can do
+`X()` when `X` denotes an object that has a `call` method).
+
+The constructors named `C.named` and `D.named` are treated similarly except
+that they are named. They can be invoked using expressions like
+`X.named(42)` when `X` is a type variable that
+`static extends Creation<X>`.
+
+When a class, mixin, mixin class, or enum declaration has a
+`static implements I` clause, the metaobject class for said type will have
+an `implements I` clause, and it is implemented by generating forwarding
+instance members for each of the members of the interface of `I`.
+
+It is also possible to use a `static extends T` clause, in which case the
+metaobject class will have an `extends T` clause. This implies that the
+metaobject class can inherit behaviors from `T` (and possibly implement
+others as forwarders, with members which are not implemented otherwise).
```

Comment:
No `static with`? Could be put into the `extends`, as `static extends Foo with Mixin`.